SAI: less hardcoded

eurofun opened the pull request:
The imx sai driver currently uses a lot of hardcoded parameters and of 6 sais only 2 are defined in the firmware.
I made changes to have sai2 and 3 work with 8ch each (pb & capture). Tested on my custom imx8mp board, works fine.

I use the btsco codec driver with two dai links in a single simple-audio-card. The linux-imx 6.1 kernel has a problem with that 
and I will post details about the problem and my workaround later on.

Is there any reason I can't grasp why not all sais are defined? Is it because that would use a little bit more memory? I'm good with sai2 and 3...

I don't have a NXP evkit so I can't test if the changes break something. One candidate for such behaviour
would be the fifo packing settings: to enable pack8/pack16 and adapt the 'fifo priming' to always write one full frame, but at least one fifo word. I'm not fully understanding why that priming is needed in the first place

Some of the commits are more questions (like on the sdma watermark level) and I'm aware that the comment style likely has to be adapted (used some // comments) among some other things - thus marked as draft.

sofci commented:
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

paulstelian97 commented on src/platform/imx8m/lib/dai.c:
```diff
@@ -34,6 +39,28 @@ static SHARED_DATA struct dai sai[] = {
+			.depth = 128, /* in 4 bytes words */
+			.watermark = 64, /* half the depth */
+			/* Handshake is SDMA hardware event */
+			.handshake = 3,
```
Did you verify these handshake numbers? You can find the SDMA hardware event numbers in the i.MX8MP reference manual. The naming "handshake" is an unfortunate consequence on how SOF defined the structure in the first place, I have used the "handshake" to provide information specific to the DMA driver (in the case of SDMA, the hardware event which is to be used)

paulstelian97 commented on src/drivers/imx/sdma.c:
```diff
@@ -691,6 +691,7 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
 		 * For MEM2MEM the source is stored in buf_addr and destination
 		 * is in buf_xaddr.
 		 */
+		tr_dbg(&sdma_tr, "SDMA: elem index %d dir %d", i, config->direction);
```
Before you remove this pull request from draft, please remove these prints. Debug prints can be added locally, however unless there's a _very_ clear case where they're useful they should generally not be submitted upstream.

eurofun replied:
I would have removed them together with the whole sdma commit for a non-draft pull request. I just left it in here as a convenient way to ask my question about the watermark level.

paulstelian97 commented on src/drivers/imx/sdma.c:
```diff
@@ -750,8 +753,18 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
+	tr_info(&sdma_tr, "SDMA: burst_elems %d width %d array count %d",
+		config->burst_elems, width, config->elem_array.count);
+	tr_info(&sdma_tr, "SDMA: src_width %d dest_width %d watermark %d",
+		config->src_width, config->dest_width, watermark);
```
Please do not add unconditional tr_info prints that run every period. That will cause a lot of spam in the logs, preventing us from seeing actually important messages. At a minimum downgrade these to tr_dbg prints, though I don't really see the value of having them in the upstream code in the first place.

I suppose you can keep these prints while the pull request is still in draft mode, if they help you debug. But remember to remove them.

eurofun replied:
These two only run on setup, at least that's what sof-logger is showing me. I made them info because dbg is not so helpful because then the spamming you mentioned really starts. I will remove them.

paulstelian97 reviewed (commented):
Please check these comments. I did not provide a _full_ review as of now, these changes definitely are interesting, but I'm not fully convinced of them yet.

gkbldcig commented:
Can one of the admins verify this patch?

paulstelian97 commented:
> The imx sai driver currently uses a lot of hardcoded parameters and of 6 sais only 2 are defined in the firmware. I made changes to have sai2 and 3 work with 8ch each (pb & capture). Tested on my custom imx8mp board, works fine.

The only reason we only defined two SAIs is because we only used those. Feel free to add more as you use them.

> 
> I don't have a NXP evkit so I can't test if the changes break something. One candidate for such behaviour would be the fifo packing settings: to enable pack8/pack16 and adapt the 'fifo priming' to always write one full frame, but at least one fifo word. I'm not fully understanding why that priming is needed in the first place

I don't recall exactly why that was added (it wasn't there in my first revision of the driver), but I do have a vague memory of it being done to resolve certain issues due to the timing on when the DMA and DAI drivers are started. If you don't feed those zero bytes the SAI may have a chance to underrun before the DMA is ever started and that can prevent playback from starting. See commit 07da0a8145 (taken from "git blame") for more details on that.

On the watermark question, again "git blame" points to why it became as it is today -- an experimentally found value. See commit 12fdd638de for more details.

eurofun commented:
Thanks for reminding me to use git blame, somehow didn't get around doing it here...
I was expecting something like that for the watermark level and the fifo priming. It works nicely for my dual 8ch setup so something must be right - but certainly a good place for further experiments, time permitting.

What is the easiest way to tell a fw build is using zephyr or xtos? After I updated the main branch today logger started to say
` src/audio/dai-legacy.c:655  INFO dai_zephyr_config_prepare(), channel = 0`
Is that a sure sign that a switch to zephyr happened? 

The failing of checkpatch is a false-positive in my opinion unless forming the plural of SAI as SAIs is really wrong. Will work around that...

paulstelian97 commented:
> What is the easiest way to tell a fw build is using zephyr or xtos? After I updated the main branch today logger started to say ` src/audio/dai-legacy.c:655 INFO dai_zephyr_config_prepare(), channel = 0` Is that a sure sign that a switch to zephyr happened?

I am unsure exactly how to predict whether it was Zephyr that we're using. The most visible way to see this on an already built firmware file is to see if there are any logs from the scheduler (the LL scheduler logs come from different files between Xtos and Zephyr)

As for firmware you're building yourself, it's pretty simple. If you use the `./scripts/xtensa_build_all.sh` way, it's Xtos. If you use `./scripts/xtensa_build_zephyr.py`, it's Zephyr. The output files go in different directories too dependent on which way to build it is used.

In particular your own log entry is _not_ really relevant. That function is a poorly named/compatibility shim function, as opposed to being genuinely Zephyr specific. On i.MX we still have Xtos builds (Intel has mostly, if not even entirely, done away with them) however there are plans to do a full migration to Zephyr so that we align with Intel.

dbaluta commented:
> What is the easiest way to tell a fw build is using zephyr or xtos?

At runtime you can look inside /sys/kerne/debug/sof/cc_info

dbaluta commented:
@eurofun thanks a lot for your changes! We will try to review them these days. 

Care to share with us what is your audio scenario? and what hardware are you using?

eurofun commented:
I will give SOF on Zephyr a try next week, thanks for the info!

@dbaluta The audio scenario is an ip audio gateway with 8 speaker amplifiers on one SAI and 4 codecs on the other SAI. I will check with my bosses how much further I can go into detail...
The hw are custom boards based on imx8mp smarc modules from avnet/msc (thus I'm limited to SAIs 2 and 3).

paulstelian97 commented on src/drivers/imx/sai.c:
```diff
@@ -24,7 +24,7 @@ LOG_MODULE_REGISTER(sai, CONFIG_SOF_LOG_LEVEL);
 		 0xa0, 0xa7, 0xdc, 0xa5, 0x38, 0xef, 0x81, 0xf4);
 
-DECLARE_TR_CTX(sai_tr, SOF_UUID(sai_uuid), LOG_LEVEL_INFO);
+DECLARE_TR_CTX(sai_tr, SOF_UUID(sai_uuid), LOG_LEVEL_DEBUG);
```
This looks like another one of the changes that he's only using for testing as opposed to for the actual patch. I guess it's best to skip these things (I believe I made my point quite clear in the idea that this shouldn't be in the final code) when reviewing this pull request. But I agree that this change shouldn't be in the final code.

eurofun commented:
I removed the dbg logging, added all SAIs, increased available 'runtime' heap blocks,  modified / removed some comments.

In my first comment above I talked about the kernel having problems with two dai-links in one simple-audio-card: that disappeared once I gave both DAI_CONFIGs link_id 0.

paulstelian97 reviewed (approved):
No more obvious issues I could find with your code. This still needs testing, both on NXP hardware and your own (I assume you have already done the second part).

dbaluta commented:
@eurofun can you also share with us the kernel changes? 

Will start CI with your changes tonight and get back with the results. So we appreciate your patience. 

We definitely want to merge this changes but we want to make sure that we don't break existing usecases.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -250,15 +276,15 @@ static inline int sai_set_config(struct dai *dai, struct ipc_config_dai *common_
 		val_cr2 |= REG_SAI_CR2_BCP;
-		val_cr4 |= REG_SAI_CR4_SYWD(0U);
+		val_cr4 |= REG_SAI_CR4_SYWD(1U);
 		break;
```
Not sure I understand this change. Is this a bugfix? We should add it in a separate patch with proper explanation.

eurofun replied:
I added a separate commit. Or did you mean put it into a separate pull request?

lgirdwood commented:
@dbaluta @paulstelian97 ping ?

dbaluta commented:
@LaurentiuM1234  can you have a look at this? I think we can merge this into main and have it for stable-v2.7 release.

dbaluta commented:
@eurofun sorry for getting at this so late. We will plan this for merge for stable-v2.7. As for the clocking issue, please create a separate PR.

@LaurentiuM1234  is working on moving clock control to FW once we completely switch to zephyr and deprecated xtos support.

lyakh commented on src/drivers/imx/sai.c:
```diff
@@ -33,8 +33,10 @@ static void sai_start(struct dai *dai, int direction)
+	struct sai_pdata *sai = dai_get_drvdata(dai);
 	int chan_idx = 0;
 	uint32_t xcsr = 0U;
+	int n = 0;
```
no need to initialise

eurofun replied:
Changed it.

eurofun commented:
No worries @dbaluta, I'm patient (or at least trying to be).

I'm now using RI-2020.4-linux + hifi4_mscale_v1_0_2_prod_linux.tgz  as my xtensa toolchain (latest version with xt-xcc, from there on only xt-clang):
this is why I removed the two for loop inline int declarations as they lead to a xt-xcc warning. And I removed the superfluous initialisation as suggested by lyakh). I adapted the HEAP_RUNTIME_SIZE to match what is in the nxp v2.6 stable branch as that seems ok as well. And last but not least another force push to get rid of a submodule change that I committed accidentally...

LaurentiuM1234 commented on src/drivers/imx/sai.c:
```diff
@@ -201,21 +229,20 @@ static inline int sai_set_config(struct dai *dai, struct ipc_config_dai *common_
+	 * The division value is (DIV + 1) * 2.
+	 */
+
+	uint32_t clk_div = (config->sai.mclk_rate / config->sai.bclk_rate / 2) - 1;
```
What if `MCLK` and `BCLK` have the same frequency? You'll probably have to bypass the `BCLK` divider using the `BYP` bit from xCR2.

dbaluta commented:
@lgirdwood  @dabekjakub  I see some compilation errors related to virtual heaps. Can you please check?

lgirdwood commented:
> @lgirdwood @dabekjakub I see some compilation errors related to virtual heaps. Can you please check?

Yeah, sorry that was fixed late last night. It's on HEAD so not sure why the reabse did not get it here ?

eurofun commented:
@dbaluta The checks seemingly were not re-run after my rebase onto main and the push?

lgirdwood commented:
@eurofun did you rebase on top of HEAD from today ? I can confirm I did this on another PR today and the force push fixed the GH action build failures releted to VMH.

eurofun commented:
@lgirdwood Yes, last commit is 658cb65bb5a8e12d5f914c5c378f2ea4060a2eeb. I know nothing about this CI but it seems to me yesterdays run did not completely finish yet (I'm seeing e.g. "Internal Intel CI System/merge/build Expected — Waiting for status to be reported") so no new run can start? On the other hand I'm sure there is some timeout...

dbaluta commented:
SOFCI TEST

lgirdwood commented:
> @lgirdwood can you use your super powers and forcefully merge this.

Something funny going on here, GH is showing all CI tests are running slowly. I will give it until lunch time and merge if its only vmh errors like last time.

lgirdwood commented:
> > @lgirdwood can you use your super powers and forcefully merge this.
> 
> Something funny going on here, GH is showing all CI tests are running slowly. I will give it until lunch time and merge if its only vmh errors like last time.

Looks like normal service now resumed and all original failed tests passing. I'll wait a few more hours so we can get a all results in, not expecting any more issues.

lgirdwood commented:
ok, jenkins/intel CI passed yesterday and rebase with VMH is passing GH actions. We can merge since Intel CI is still pending and does not test NXP code.

dbaluta commented:
Thanks a lot @eurofun  for the PR!

eurofun commented:
You're welcome @dbaluta! I'm seeing some problems (only) on Zephyr (crash if I start more than one aplay/arecord). Not sure if some other modifications I made are to blame or if it is a general problem. I will try a 'stock' setup next week.
